waku/v2/protocol/waku_message/topics/content_topic.nim

```diff
@@ -0,0 +1,89 @@
+## Waku content topics definition and namespacing utils
+##
+## See 23/WAKU2-TOPICS RFC: https://rfc.vac.dev/spec/23/
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/strutils,
+  stew/results
+import
+  ./parsing
+
+export parsing
+
+
+## Content topic
+
+type ContentTopic* = string
+
+const DefaultContentTopic* = ContentTopic("/waku/2/default-content/proto")
+
+
+## Namespaced content topic
+
+type
+  NsContentTopic* = object
+    application*: string
+    version*: string
+    name*: string
+    encoding*: string
+
+proc init*(T: type NsContentTopic, application, version, name, encoding: string): T =
+  NsContentTopic(
+    application: application,
+    version: version,
+    name: name,
+    encoding: encoding
+  )
+
+
+# Serialization
+
+proc `$`*(topic: NsContentTopic): string =
+  ## Returns a string representation of a namespaced topic
+  ## in the format `/<application>/<version>/<topic-name>/<encoding>`
+  "/" & topic.application & "/" & topic.version & "/" & topic.name & "/" & topic.encoding
+
+
+# Deserialization
+
+proc parse*(T: type NsContentTopic, topic: ContentTopic|string): ParsingResult[NsContentTopic] =
+  ## Splits a namespaced topic string into its constituent parts.
+  ## The topic string has to be in the format `/<application>/<version>/<topic-name>/<encoding>`
+
+  if not topic.startsWith("/"):
+    return err(ParsingError.invalidFormat("topic must start with slash"))
+
+  let parts = topic[1..<topic.len].split("/")
+  if parts.len != 4:
+    return err(ParsingError.invalidFormat("invalid topic structure"))
+
+
+  let app = parts[0]
+  if app.len == 0:
+    return err(ParsingError.missingPart("appplication"))
+
+  let ver = parts[1]
+  if ver.len == 0:
+    return err(ParsingError.missingPart("version"))
+
+  let name = parts[2]
+  if name.len == 0:
+    return err(ParsingError.missingPart("topic-name"))
+
+  let enc = parts[3]
+  if enc.len == 0:
+    return err(ParsingError.missingPart("encoding"))
+
+
+  ok(NsContentTopic.init(app, ver, name, enc))
+
+
+# Content topic compatibility
+
+converter toContentTopic*(topic: NsContentTopic): ContentTopic =
```

Comment:
I like and agree with explicitness. I also considered adding a `distinct string` to the topics, but there is a small issue: 

The `ContentTopic` and the `PubsubTopic` types are part of the "core" Waku types. If we want it to be explicit, we have to change all the places where we accept an argument of type `ContentTopic` to `ContentTopic|NsContentTopic`. This will lead to a massive PR with little gain.

In the end, I opted for this small `converter` to support an implicit "one-way" conversion (since, in the opposite direction, the conversion is fallible).

```Nim

let pubsubTopic = NsPubsubTopic.staticSharding(0, 2)

proc doSomething(topic: PubsubTopic) =
	## Do something

doSomething(pubsubTopic)  # This should compile
````